### PR TITLE
Implement take for StreamArrayReader

### PR DESCRIPTION
**Pull request by jdcasale**

### CODE REVIEW

File: vortex-ipc/Cargo.toml [hunk 1]
```diff
@@ -31,6 +31,8 @@ walkdir = { workspace = true }
 criterion = { workspace = true }
 rand = { workspace = true }
 simplelog = { workspace = true }
+vortex-fastlanes = { path = "../vortex-fastlanes" }
```
Comment (gatesn) [inline vortex-ipc/Cargo.toml:34]: Can you move these into [dev-dependencies]? Same actually for zig-zag's dep on fastlanes.

File: vortex-ipc/benches/ipc_take.rs [hunk 2]
```diff
@@ -22,15 +22,17 @@ fn ipc_take(c: &mut Criterion) {
...
 
     c.bench_function("take_view", |b| {
         b.iter(|| {
             let mut cursor = Cursor::new(&buffer);
             let mut reader = StreamReader::try_new(&mut cursor).unwrap();
             let mut array_reader = reader.next().unwrap().unwrap();
-            let array_view = array_reader.next().unwrap().unwrap();
-            black_box(take(&array_view, &indices))
+            while let Some(array_chunk) = array_reader.next().unwrap() {
+                black_box(take(&array_chunk, &indices).unwrap());
```
Comment (gatesn) [inline vortex-ipc/benches/ipc_take.rs]: You're not benchmarking your own take function?

File: vortex-ipc/src/reader.rs [hunk 3]
```diff
@@ -100,6 +108,7 @@ impl<R: Read> FallibleLendingIterator for StreamReader<R> {
             dtype,
             buffers: vec![],
             column_msg_buffer: vec![],
+            byte_offset: 0,
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Is it really measured in bytes?

File: vortex-ipc/src/reader.rs [hunk 4]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+    byte_offset: usize,
 }
 
 impl<'a, R: Read> StreamArrayReader<'a, R> {
     pub fn dtype(&self) -> &DType {
         &self.dtype
     }
 
-    pub fn take(&self, indices: &Array<'_>) -> VortexResult<OwnedArray> {
+    pub fn take(&mut self, indices: &Array<'_>) -> VortexResult<OwnedArray> {
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: You should take `self`, similar to regular iterator take fn, since this array reader isn't really usable after this function has run.

Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Ideally we would ideally rewrite this to return an iterator of OwnedArray. Maybe leave a comment saying that

File: vortex-ipc/src/reader.rs [hunk 5]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
         }
-        todo!()
+
+        if self.byte_offset != 0 {
+            vortex_bail!("Stream has already been (at least partially) consumed")
+        }
+
+        let mut chunks = Vec::new();
+
+        let mut offset = self.byte_offset;
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: `row_offset` is more explicit

File: vortex-ipc/src/reader.rs [hunk 6]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+        let idx_len = indices.len();
+        let mut found_so_far = 0;
+
+        // Continue reading batches from the stream until we either run out or find all indices
+        while let Some(batch) = self.next()? {
+            let left =
+                search_sorted::<u64>(indices, offset as u64, SearchSortedSide::Left)?.to_index();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why cast everything to u64? usize should work

File: vortex-ipc/src/reader.rs [hunk 7]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            let right = search_sorted::<u64>(
+                indices,
+                offset as u64 + batch.len() as u64,
+                SearchSortedSide::Left,
+            )?
+            .to_index();
+
+            // TODO(@jdcasale): replace this with compute scalar_sum when we've added it
+            let indices_for_batch = slice(indices, left, right)?.flatten_primitive()?;
+            let ptype = indices.dtype().try_into()?;
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: You just made indices_for_batch a PrimitiveArray, it has a ptype

File: vortex-ipc/src/reader.rs [hunk 8]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                search_sorted::<u64>(indices, offset as u64, SearchSortedSide::Left)?.to_index();
+            let right = search_sorted::<u64>(
+                indices,
+                offset as u64 + batch.len() as u64,
+                SearchSortedSide::Left,
+            )?
+            .to_index();
+
+            // TODO(@jdcasale): replace this with compute scalar_sum when we've added it
+            let indices_for_batch = slice(indices, left, right)?.flatten_primitive()?;
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:206]: Should skip the whole batch if left == right.

File: vortex-ipc/src/reader.rs [hunk 9]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            let ptype = indices.dtype().try_into()?;
+            let shifted = match_each_unsigned_integer_ptype!(ptype, |$P| {
+                indices_for_batch
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Small thing, but you could construct the PrimitiveArray inside this match statement. If the indices are u8, this means they can remain u8 even after shifting.

File: vortex-ipc/src/reader.rs [hunk 10]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                indices,
+                offset as u64 + batch.len() as u64,
+                SearchSortedSide::Left,
+            )?
+            .to_index();
+
+            // TODO(@jdcasale): replace this with compute scalar_sum when we've added it
+            let indices_for_batch = slice(indices, left, right)?.flatten_primitive()?;
+            let ptype = indices.dtype().try_into()?;
+            let shifted = match_each_unsigned_integer_ptype!(ptype, |$P| {
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why do you assume indices are unsigned integers? This will panic if they're not.

File: vortex-ipc/src/reader.rs [hunk 11]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                indices_for_batch
+                    .typed_data::<$P>()
+                    .iter()
+                    .map(|&idx| {
+                        idx  as u64 - offset as u64
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Surprised `cargo fit` allows the double space

File: vortex-ipc/src/reader.rs [hunk 12]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                    })
+                    .collect_vec()
+            });
+
+            let shifted_arr = PrimitiveArray::from(shifted).to_array().to_static();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why to_static? You just created it from heap allocated vec, it's already static.

File: vortex-ipc/src/reader.rs [hunk 13]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            let from_current_batch = take(&batch, &shifted_arr)?;
+            println!("from_current_batch: {:?}", from_current_batch);
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Remove

File: vortex-ipc/src/reader.rs [hunk 14]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            found_so_far += from_current_batch.len();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: It's a little confusing to take the length from this array. Vs the length of the indices_for_batch. You could assert_eq the lengths as a sanity check.

File: vortex-ipc/src/reader.rs [hunk 15]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            chunks.push(from_current_batch);
+            offset += batch.len();
+            if found_so_far >= indices.len() {
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: A better check is `left == indices.len()` above, then you don't need to keep track of anything.

File: vortex-ipc/src/reader.rs [hunk 16]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                break;
+            }
+        }
+        self.byte_offset = offset;
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: The whole point of `self.ROW_offset` is to check whether the iterator has been consumed. You should be updating this in `next()` not here.

File: vortex-ipc/src/reader.rs [hunk 17]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+        if found_so_far < idx_len {
+            return Err(vortex_err!("EOF encountered before finding all indices"));
+        }
+
+        let result = if chunks.len() == 1 {
+            chunks.first().unwrap().to_static()
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why to_static?

File: vortex-ipc/src/reader.rs [hunk 18]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+        } else {
+            ChunkedArray::try_new(chunks, self.dtype.clone())?
+                .into_array()
+                .to_static()
+        };
+        Ok(OwnedArray::from(result))
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Chunks is already a `Vec<OwnedArray>` isn't it?

File: vortex-ipc/src/reader.rs [hunk 19]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+        if found_so_far < idx_len {
+            return Err(vortex_err!("EOF encountered before finding all indices"));
+        }
+
+        let result = if chunks.len() == 1 {
+            chunks.first().unwrap().to_static()
+        } else {
+            ChunkedArray::try_new(chunks, self.dtype.clone())?
+                .into_array()
+                .to_static()
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why to_static?

File: vortex-ipc/src/reader.rs [hunk 20]
```diff
@@ -193,3 +260,158 @@ impl<'iter, R: Read> FallibleLendingIterator for StreamArrayReader<'iter, R> {
...
+    use vortex_fastlanes::BitPackedEncoding;
+    use vortex_schema::{DType, IntWidth, Nullability, Signedness};
+
+    use crate::iter::FallibleLendingIterator;
+    use crate::reader::StreamReader;
+    use crate::writer::StreamWriter;
+
+    #[test]
+    fn test_write_read_primitive() {
+        let data = PrimitiveArray::from((0i32..3_000_000).rev().collect_vec()).into_array();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:426]: The rev is really confusing, didn't see it at first!

File: vortex-ipc/src/reader.rs [hunk 21]
```diff
@@ -193,3 +260,158 @@ impl<'iter, R: Read> FallibleLendingIterator for StreamArrayReader<'iter, R> {
...
+        test_write_read(
+            packed,
+            vec![2999989i64, 2999988, 2999987, 2999986, 2899999, 0],
+            PrimitiveEncoding.id(),
+        );
+    }
+
+    #[test]
+    fn test_write_read_chunked() {
+        let indices = PrimitiveArray::from(vec![10u32, 11, 12, 13, 100_000, 2_999_999, 3_000_000])
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: If I set indices == `10u32, 11, 12, 13, 100_000, 2_999_999, 2_999_999, 3_000_000` I get `[2999989, 2999988, 2999987, 2999986, 2899999, 0, 0]`

File: vortex-ipc/src/reader.rs [hunk 22]
```diff
@@ -111,22 +120,80 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+        let mut found_so_far = 0;
+
+        // Continue reading batches from the stream until we either run out or find all indices
+        while let Some(batch) = self.next()? {
+            let left =
+                search_sorted::<u64>(indices, offset as u64, SearchSortedSide::Left)?.to_index();
+            let right = search_sorted::<u64>(
+                indices,
+                offset as u64 + batch.len() as u64,
+                SearchSortedSide::Left,
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Always write tests that cover your boundary conditions

File: vortex-ipc/src/reader.rs [hunk 23]
```diff
@@ -123,22 +132,79 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+            )?
+            .to_index();
+
+            // TODO(@jdcasale): replace this with compute scalar_sum when we've added it
+            let indices_for_batch = slice(indices, left, right)?.flatten_primitive()?;
+            let ptype = indices.dtype().try_into()?;
+            let shifted = match_each_unsigned_integer_ptype!(ptype, |$P| {
+                indices_for_batch
+                    .typed_data::<$P>()
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:209]: Don't forget the primitive array may have a validity buffer.

File: vortex-ipc/src/reader.rs [hunk 24]
```diff
@@ -123,22 +132,79 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                        idx  as u64 - offset as u64
+                    })
+                    .collect_vec()
+            });
+
+            let shifted_arr = PrimitiveArray::from(shifted).to_array().to_static();
+            let from_current_batch = take(&batch, &shifted_arr)?;
+            found_so_far += from_current_batch.len();
+            chunks.push(from_current_batch);
+            offset += batch.len();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: `self.offset` should already have this information no?

Comment (gatesn) [reply vortex-ipc/src/reader.rs]: Ah no, you don't update it in the right place

File: vortex-ipc/src/reader.rs [hunk 25]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_write_read_bitpacked() {
+        let uncompressed = PrimitiveArray::from((0i64..3_000_000).rev().collect_vec());
+
+        let cfg = CompressConfig::new().with_enabled([&BitPackedEncoding as EncodingRef]);
+        let ctx = CompressCtx::new(Arc::new(cfg));
+
+        let packed = BitPackedEncoding {}
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Add a `BitPackedArray::encode` instead of reaching into the encoding plugin endpoint.

File: vortex-ipc/src/reader.rs [hunk 26]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+        let data = PrimitiveArray::from((0i32..3_000_000).rev().collect_vec()).into_array();
+        let data2 =
+            PrimitiveArray::from((3_000_000i32..6_000_000).rev().collect_vec()).into_array();
+        let chunked = ChunkedArray::try_new(
+            vec![data, data2],
+            DType::Int(IntWidth::_32, Signedness::Signed, Nullability::NonNullable),
+        )
+        .unwrap()
+        .into_array();
+        let mut buffer = vec![];
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:514]: Can we have some line breaks? This is like writing without paragraphs 😅

File: vortex-ipc/src/reader.rs [hunk 27]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+        {
+            let mut cursor = Cursor::new(&mut buffer);
+            {
+                let mut writer =
+                    StreamWriter::try_new(&mut cursor, SerdeContext::default()).unwrap();
+                writer.write_array(&chunked).unwrap();
+                writer.write_array(&chunked).unwrap();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Why write the array twice?

File: vortex-ipc/src/reader.rs [hunk 28]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+        assert_eq!(array.encoding().id(), ChunkedEncoding.id());
+        let bind = ChunkedArray::try_from(array).unwrap();
+        let mut chunked_out = bind.chunks();
+        assert_eq!(
+            chunked_out
+                .next()
+                .unwrap()
+                .into_primitive()
+                .typed_data::<u32>()
+                .to_vec(),
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: You can compare slices without copying it into a vec

File: vortex-ipc/src/reader.rs [hunk 29]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+        let indices =
+            PrimitiveArray::from(vec![10u32, 11, 12, 13, 100_000, 2_999_999]).into_array();
+        let mut buffer = vec![];
+        {
+            let mut cursor = Cursor::new(&mut buffer);
+            {
+                let mut writer =
+                    StreamWriter::try_new(&mut cursor, SerdeContext::default()).unwrap();
+                writer.write_array(&data).unwrap();
+                writer.write_array(&data).unwrap();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:571]: Same again, why write twice?

File: vortex-ipc/src/reader.rs [hunk 30]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+                .unwrap()
+                .into_primitive()
+                .typed_data::<u32>()
+                .to_vec(),
+            vec![5999999]
+        );
+    }
+    fn test_write_read_inner<T: NativePType>(
+        data: Array,
+        expected: Vec<T>,
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Typically would take `&[T]` here

File: vortex-ipc/src/reader.rs [hunk 31]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+    #[test]
+    fn test_write_read_chunked() {
+        let indices = PrimitiveArray::from(vec![10u32, 11, 12, 13, 100_000, 2_999_999, 3_000_000])
+            .into_array();
+        let data = PrimitiveArray::from((0i32..3_000_000).rev().collect_vec()).into_array();
+        let data2 =
+            PrimitiveArray::from((3_000_000i32..6_000_000).rev().collect_vec()).into_array();
+        let chunked = ChunkedArray::try_new(
+            vec![data, data2],
+            DType::Int(IntWidth::_32, Signedness::Signed, Nullability::NonNullable),
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Use `data.dtype()`

File: vortex-ipc/src/reader.rs [hunk 32]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+                writer.write_array(&chunked).unwrap();
+            }
+        }
+
+        let mut cursor = Cursor::new(&buffer);
+        let mut reader = StreamReader::try_new(&mut cursor).unwrap();
+        let mut array_reader = reader.next().unwrap().unwrap();
+        let array = array_reader.take(&indices).unwrap();
+        assert_eq!(array.encoding().id(), ChunkedEncoding.id());
+        let bind = ChunkedArray::try_from(array).unwrap();
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: bind is a bad name that IntelliJ generates for you

File: vortex-ipc/src/reader.rs [hunk 33]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+        let mut chunked_out = bind.chunks();
+        assert_eq!(
+            chunked_out
+                .next()
+                .unwrap()
+                .into_primitive()
+                .typed_data::<u32>()
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Huh, this shouldn't work. I think during the refactor we dropped the assertion that `T::PTYPE == array.ptype()` in typed_data and scalar_buffer. Could you add it back, then fix this to be i32?

File: vortex-ipc/src/reader.rs [hunk 34]
```diff
@@ -316,4 +391,138 @@ mod tests {
...
+                .to_vec(),
+            vec![2999989, 2999988, 2999987, 2999986, 2899999, 0]
+        );
+        assert_eq!(
+            chunked_out
+                .next()
+                .unwrap()
+                .into_primitive()
+                .typed_data::<u32>()
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: Likewise this should be i32

**Review by gatesn — CHANGES_REQUESTED** [review]

File: vortex-ipc/src/reader.rs [hunk 35]
```diff
@@ -123,22 +132,79 @@ pub struct StreamArrayReader<'a, R: Read> {
...
+                    .collect_vec()
+            });
+
+            let shifted_arr = PrimitiveArray::from(shifted).to_array().to_static();
+            let from_current_batch = take(&batch, &shifted_arr)?;
+            found_so_far += from_current_batch.len();
+            chunks.push(from_current_batch);
+            offset += batch.len();
+            if found_so_far >= indices.len() {
+                break;
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs]: You can't really just break out here since now the stream will be left somewhere in the middle. You have to consume the rest of the array and throw it away so that the next call `StreamReader.next()` is correct.

File: vortex-ipc/src/reader.rs [hunk 36]
```diff
@@ -275,11 +341,20 @@ impl StreamMessageReader {
 #[cfg(test)]
 mod tests {
```
Comment (gatesn) [inline vortex-ipc/src/reader.rs:372]: Can you add a test with a stream containing two arrays that if you call take(...) on the first one, you can then successfully read the second one. I think right now you leave the first array half-consumed if there are no indices in the final chunk.

File: vortex-ipc/src/reader.rs (hunk 21 above)
Comment (jdcasale) [reply vortex-ipc/src/reader.rs]: Resolved offline, this ^ behavior is actually correct

File: vortex-ipc/Cargo.toml (hunk 1 above)
Comment (jdcasale) [reply vortex-ipc/Cargo.toml:34]: Screenshot 2024-04-26 at 1 19 01 PM

I think I'm misunderstanding something -- this is already in dev-dependencies?

File: vortex-ipc/src/reader.rs (hunk 20 above)
Comment (jdcasale) [reply vortex-ipc/src/reader.rs:426]: will leave a comment

File: vortex-ipc/src/reader.rs (hunk 17 above)
Comment (jdcasale) [reply vortex-ipc/src/reader.rs]: I should have reached for clone() here

File: vortex-ipc/src/reader.rs (hunk 4 above)
Comment (jdcasale) [reply vortex-ipc/src/reader.rs:143]: The compiler complains about calling `self.next()` if I don't borrow as mutable here. Maybe you've fixed this in the upstream iterator branch?

File: vortex-ipc/benches/ipc_take.rs (hunk 2 above)
Comment (jdcasale) [reply vortex-ipc/benches/ipc_take.rs]: Yeah I think this branch was just outdated when you saw it -- fixed now.

File: vortex-array/src/ptype.rs [hunk 37]
```diff
@@ -137,7 +137,23 @@ macro_rules! match_each_integer_ptype {
         }
     })
 }
+
+#[macro_export]
+macro_rules! match_each_unsigned_integer_ptype {
```
Comment (gatesn) [inline vortex-array/src/ptype.rs]: I don't think you need this now you support signed integer ptypes.

**Review by gatesn — APPROVED** [review]
